### PR TITLE
refactor: better resource cleanup handling

### PR DESCRIPTION
**Pull request by eddycharly**
**What this PR does / why we need it**:

This PR improves resource cleanup handling.
- moves cleanup code closer to the resource creation code
- reuse discovery information used at resource creation time to make creation/cleanup more robust
- only cleanup a resource if kuttl created it (don't register cleanup if it was an update)

### CODE REVIEW

File: pkg/test/case.go
Comment (eddycharly) [reply pkg/test/case.go:347]: Not sure it would make things better, currently this is not overridable in a `Step`, this makes sense to me that it is the responsibility of the caller to pass the flag.
I don't have a strong opinion though, if you prefer the field based approach I will use that.
WDYT ?

File: pkg/test/step.go
Comment (eddycharly) [reply pkg/test/step.go]: Fixed

File: pkg/test/case.go
Comment (eddycharly) [reply pkg/test/case.go]: Switched to a struct field.

File: pkg/test/step.go
Comment (iblancasa) [inline pkg/test/step.go:204]: Please, remove this `TODO` and create an issue instead

**Review by iblancasa — CHANGES_REQUESTED** [review]

File: pkg/test/step.go
Comment (eddycharly) [reply pkg/test/step.go:204]: https://github.com/kudobuilder/kuttl/issues/420

Comment (eddycharly) [reply pkg/test/step.go]: I removed the `TODO`

**Review by iblancasa — APPROVED** [review]

**Review by erikgb — APPROVED** [review]
\lgtm

**Review by kensipe — CHANGES_REQUESTED** [review]
integration tests need updating for completion

```
./bin/golangci-lint --timeout 5m run --build-tags integration
pkg/test/step_integration_test.go:443:28: not enough arguments in call to step.Run
	have (string)
	want (*"testing".T, string) (typecheck)
	errs := step.Run(namespace)
	                          ^
pkg/test/step_integration_test.go:461:33: not enough arguments in call to step.Run
	have (string)
	want (*"testing".T, string) (typecheck)
	errors := step.Run("irrelevant")
	                               ^
pkg/test/step_integration_test.go:479:33: not enough arguments in call to step.Run
	have (string)
	want (*"testing".T, string) (typecheck)
	errors := step.Run("irrelevant")
```

**Comment by kensipe** [comment]
for awareness... I commonly run the following as a starting point of review:  `make lint test integration-test`

`make lint` is failing at this point.   Likely missed based on the `integration` build tag

**Comment by eddycharly** [comment]
I will fix this tomorrow. How about running this a gh workflow ?

**Comment by kensipe** [comment]
yeah... when I noticed the passing CI... I had the same thought.  I will look into that.

**Comment by kensipe** [comment]
well... now I remember some issues we had in the past but haven't been resolved.   For some reason CircleCI wouldn't run PRs from non-committers.   So the full CI isn't run (just the DCO).  I worked with circle to help resolve that but didn't get it resolved.   We were looking to move to Github Actions in part to solve that... but it also reduces the number of CI dependencies.   Probably need to refocus on that.

**Comment by kensipe** [comment]
that's what https://github.com/kudobuilder/kuttl/pull/337 was intended to solve.

**Review by kensipe — APPROVED** [review]
thanks @eddycharly .  I pushed the integration test refactors.  I'll merge after good CI runs.   Thanks for the contribution!

**Pull request merged**
